### PR TITLE
 nixos/realm: init 

### LINKED ISSUE #284277: Package request: realm
**Project description**
A simple, high performance relay server written in rust.

**Metadata**

* homepage URL: https://github.com/zhboner/realm
* source URL: https://github.com/zhboner/realm
* license: mit
* platforms: Linux, darwin

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by ocfox**
## Description of changes
Close #284277, a simple, high performance relay server written in rust.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/networking/realm.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        default = { };
+        description = mdDoc ''
+          The realm configuration, see <https://github.com/zhboner/realm#overview> for documentation.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.packages = [ cfg.package ];
```
Comment (NickCao) [inline nixos/modules/services/networking/realm.nix:34]: The packages does not seem to provide a systemd unit file?

File: nixos/modules/services/networking/realm.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    systemd.services.realm = {
+      serviceConfig = {
```
Comment (NickCao) [inline nixos/modules/services/networking/realm.nix:36]: Could use some hardening, `DynamicUsers` etc.

File: nixos/tests/realm.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  meta = {
+    maintainers = with lib.maintainers; [ ocfox ];
+  };
+
+  nodes.machine = { pkgs, ... }: {
+    services.nginx = {
+      enable = true;
+      statusPage = true;
+    };
+    # realm need DNS resolv server to run or use config.dns.nameserver
```
Comment (NickCao) [inline nixos/tests/realm.nix:13]: Shall this be enforced in the module?

File: pkgs/by-name/re/realm/package.nix [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    hash = "sha256-RcaFWR23FlMNuyTQ/7YbHm/k0AhkfsLX+qA10PW1Jfg=";
+  };
+
+  cargoHash = "sha256-BBZZNJYBW7hLYvIEzyTZtTKHlxF/7puFuRPJrmOjV/8=";
+
+  buildInputs = lib.optionals stdenv.isDarwin [
+    darwin.apple_sdk.frameworks.Security
+  ];
+
+  RUSTC_BOOTSTRAP = 1;
```
Comment (NickCao) [inline pkgs/by-name/re/realm/package.nix:27]:
```suggestion
  env.RUSTC_BOOTSTRAP = 1;
```

File: pkgs/by-name/re/realm/package.nix [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  passthru = {
+    updateScript = nix-update-script { };
+    tests = { inherit (nixosTests) realm; };
+  };
+
+
+  meta = with lib; {
+    description = "A network relay tool";
```
Comment (NickCao) [inline pkgs/by-name/re/realm/package.nix:36]: Be more informative.

File: nixos/tests/realm.nix (hunk 3 above)
Comment (ocfox) [reply nixos/tests/realm.nix:13]: It simply reads `resolv.conf`, should not normally encounter problems.

File: pkgs/by-name/re/realm/package.nix [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+  env.RUSTC_BOOTSTRAP = 1;
+
+  passthru = {
+    updateScript = nix-update-script { };
+    tests = { inherit (nixosTests) realm; };
+  };
+
+
+  meta = with lib; {
```
Comment (SuperSandro2000) [inline pkgs/by-name/re/realm/package.nix:35]:
```suggestion
  };

  meta = with lib; {
```

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by ocfox** [comment]
Thank you, I almost forgot about this PR.

**Pull request merged**